src/constraints.jl

```diff
@@ -0,0 +1,555 @@
+    JuMP.is_valid(model, con.lvref) || error("Logical variable belongs to a different model.")
+    if !haskey(_indicator_to_constraints(model), con.lvref)
+        _indicator_to_constraints(model)[con.lvref] = Vector{Union{DisjunctConstraintRef, DisjunctionRef}}()
+    end
+    push!(_indicator_to_constraints(model)[con.lvref], cref)
+    return
+end
+
+"""
+    JuMP.add_constraint(
+        model::JuMP.Model,
+        con::_DisjunctConstraint,
+        name::String = ""
+    )::DisjunctConstraintRef
+
+Extend `JuMP.add_constraint` to add a [`_DisjunctConstraint`](@ref) to a [`GDPModel`](@ref). 
+The constraint is added to the `GDPData` in the `.ext` dictionary of the `GDPModel`.
+"""
+function JuMP.add_constraint(
+    model::JuMP.Model,
+    con::_DisjunctConstraint,
+    name::String = ""
+)
+    is_gdp_model(model) || error("Can only add disjunct constraints to `GDPModel`s.")
+    data = ConstraintData(con.constr, name)
+    idx = _MOIUC.add_item(gdp_data(model).disjunct_constraints, data)
+    _add_indicator_var(con, DisjunctConstraintRef(model, idx), model)
+    return DisjunctConstraintRef(model, idx)
+end
+
+################################################################################
+#                              DISJUNCTIONS
+################################################################################
+function _process_structure(_error, s::Vector{LogicalVariableRef}, model::JuMP.Model)
+    allunique(s) ||_error("Not all the logical indicator variables are unique.")
+    for lvref in s
+        if !JuMP.is_valid(model, lvref)
+            _error("`$lvref` is not a valid logical variable reference.")
+        elseif !haskey(_indicator_to_constraints(model), lvref)
+            _error("`$lvref` is not associated with any constraints.")
+        end
+    end
+    return s
+end
+
+# fallback
+function _process_structure(_error, s, model::JuMP.Model)
+    _error("Unrecognized disjunction input structure.") # TODO add details on proper syntax
+end
+
+# Write the main function for creating disjunctions that is macro friendly
+function _create_disjunction(
+    _error::Function,
+    model::JuMP.Model, # TODO: generalize to AbstractModel
+    structure::Vector,
+    name::String,
+    nested::Bool
+)
+    is_gdp_model(model) || error("Can only add disjunctions to `GDPModel`s.")
+
+    # build the disjunction
+    indicators = _process_structure(_error, structure, model)
+    disjunction = Disjunction(indicators, nested)
+
+    # add it to the model
+    disjunction_data = ConstraintData(disjunction, name)
+    idx = _MOIUC.add_item(_disjunctions(model), disjunction_data)
+
+    _set_ready_to_optimize(model, false)
+    return DisjunctionRef(model, idx)
+end
+
+# Disjunction build for unnested disjunctions
+function _disjunction(
+    _error::Function,
+    model::JuMP.Model, # TODO: generalize to AbstractModel
+    structure::Vector,
+    name::String
+)
+    _create_disjunction(_error, model, structure, name, false)
```

Comment:
This should use `return` as per the style guide
```suggestion
    return _create_disjunction(_error, model, structure, name, false)
```